### ISSUE TITLE
Design issues for data validation preprocessing program

### ISSUE THREAD

**Issue by pavlis** [wangyinz/mspass#59]
This new issue comes up because I think we will likely need it soon.   It is prudent to get started on this key component, in my opinion, for MsPASS to be successful as a processing framework.

Background on this is a well-established idea in seismic reflection data processing.  Every reflection processing system I know of has a validation stage before initiating a processing workflow.   What “validation” means is variable, but the whole point is that it is essential to eliminate as many detectable errors or issues as possible before starting a workflow that could run for days.   

Things I know of that can and should be adapted to MsPASS are:

1.  Verify the integrity of the input data.   Make sure it exists and passes some basic sanity checks.   There are two elements to this problem:  a) validate metadata and b) sanity checks on waveforms. This will require development of one or more standalone programs.
2.  Check the syntax of the job control language.  I our case this is python and there are tools available to validate any syntax errors before starting a job.  Presumably this is just making recommendations on tools for validating syntax, but we may want something to validate calls for MsPASS algorithms.
3.  Related but different is validating the parameters passed to processing steps. Validating arguments is the classic version of this, but modern algorithms (particularly things written in a OOP language like python) can have much more complicated input (e.g. a python dict or an AntelopePf).  A generic tester for input parameter definitions will be needed.  Some thought is needed for the design of this, particularly since a long term design should be conscious of the likely desire for a gui-based method for constructing input parameteric input.  
4.   Check for availability of resources.   The limiting factor conventionally is storage space for any intermediate waveform saves.  For MsPASS we need to add some capabilities for required number of processing nodes.   That will likely be system dependent, but perhaps with the singularity container this can be done with some sophistication.

**Comment by pavlis** [comment]
An immediate thought came to me after posting that previous message.   Item 3, validating parametric input, should probably be a constraint on a valid processing function to be registered with the system.  That is, we may want to consider a formal algorithm registration system that has a few requirements:
1.  Basic documentation (e.g. docstrings)
2.  A function for parameter checking
3.   Estimates of algorithm complexity or nominal execution time (this may be better left optional). 
4.  The function itself.
5.  A test suite.

Probably don't want to be machiavellian  about it, but at least make these strong recommendation with some form of gentle scolding if any but 1 and 4 are missing.  We shouldn't be too rigid about it because in a research environment we need the flexibility of experimental code.

**Comment by wangyinz** [comment]
Verifying data and metadata is relatively straightforward. To check the syntax, there are various tools that can do it, but I guess the easiest is to just compile the python code which is supported as built-in modules of any python. 

The five requirement you listed appears more like a standard for developers to contribute to the package. It is definitely a common practice to require at least the doc and test included for any python code. 

I am not sure what is the best way to check parameters, so I just googled and little bit. Turns out that recent version of Python now supports type hints and there are tools to do static type checking. You can see look at the syntax [here](https://docs.python.org/3/library/typing.html), and there is this [mypy](https://mypy.readthedocs.io/en/stable/introduction.html) tool that can do the checking. It seems we should just adopt these.

**Comment by wangyinz** [comment]
Dug into the mypy tool a little bit more. Turns out the types we currently have will not be automatically available for type checking. We will need to generate the [stub files](https://mypy.readthedocs.io/en/latest/stubs.html#stub-files) for our C code. Seems it won't be hard, but I haven't tried yet. I tried the tool with functions written in Python that has type information annotated, it just works out of the box.

**Comment by pavlis** [comment]
Looks like you have found an excellent solution for item 3 - python function argument checking.  

I was thinking some about item 1 this morning while chasing my 10 month old grandson.   Well, more when he fell asleep on my lap and I had nothing else to occupy my mind.  I was thing most about how to verify a processing flow won't have metadata problems?   What about having an optional program that would scan the wf collection checking for the existence of required attributes?   The problem is slightly more complicated by the fact that a function may create new metadata entries that are required downstream.   That means, I think, we need a way to capture such dependencies.   

This problem is tractable for linear data flows.   It is much harder if a workflow has a conditional branch on data (e.g. handle local events differently from teleseismic events) that would run a different sequence of steps in different branches.   A way to get started is to not simply use this program as a preprocessor to suggest possible errors, but not make it a rigid requirement.  I way that because systems like ProMAX run an "init" phase that will not allow execution unless the init phase says everything is ok to proceed.  

A way to address this, I think, is to have an algorithm registration configuration file.   Any official algorithm could register itself in a yaml file with would define contraints for that algorithm.  Here is an initial list of possible items for that configuration:
1.  data input type (i.e. Seismogram, TimeSeries, or other supported data objects)
2. data output type (as above, but perhaps could be metadata only? - probably an unnecessary complexity for now) 
3.  required metadata 
4. generated metadata
5.  alters_metadata - boolean, true if it changes any metadata
6.  alters_data - boolean, true if it changes the sample data
7. input_parameter_types - list of ways the function would get initialized (yaml, AntelopePf, function arguments, maybe othes?)

A check metadata program would need to get a list of algorithms to be applied in the processing flow order.   It could scan the list and pick out all metadata required at the start of the run.   Once it had that it would be relatively easy to scan the wf collection to verify every wf entry for a specified data set has the required metadata.   

That is the rough idea I came up with this morning.  Consider this critically.

**Comment by wangyinz** [comment]
Wow, I really need to learn how to be able to think about anything when chasing my daughter around...

Yes, I think these items are pretty much the essential ones for our data validation process. I was actually more concerned of how to implement them. There are actually two different data validation needed here. One is what you described here - validation for a workflow. The other one is for the database, which should probably be called a schema validation. We need to enforce a minimum integrity of the schema to ensure the data source of any workflow is actually valid. 

I found that MongoDB has a [schema validation](https://docs.mongodb.com/manual/core/schema-validation/) feature built-in that addresses exactly the problem, so that's obviously the solution to go.

For the data validation in workflows, I was thinking that it also relates to how do we want to build our workflows. If we only want to create a standalone problem for validation, it make sense to just follow with your approach here. I found that we might use tools like [this one](https://github.com/pyeve/cerberus) to write it. If we are to integrate the validation into the Spark framework, then there might be some existing tools that we could adapt. A google search leads me to [this talk](https://databricks.com/session/apache-spark-data-validation), which shows that there are existing tools or libraries out there for similar problems. Also, if we are going to use some other workflow management tools, we might even find some that has data validation feature built-in. 

Anyway, at this stage, I think we need to determine first how a workflow looks like, and then consider developing the data validation tool so that we can build on or learn from many other applications. Maybe it is more straightforward to have the schema validation implemented now as that one is independent of the workflow.

**Comment by pavlis** [comment]
Multitasking - works for 10 month olds that don't require a lot of sophistication.  Fails miserably with any child who can communicate.   

Those are good sources that will be helpful when we do reach this point.  I think you are right that implementation is later, but I added this issue because it definitely is something we need to keep in mind to make sure as the design evolves we can make a system that is as robust as possible.

**Comment by wangyinz** [comment]
Closing this as completed for the concrete database-validation requirement.

MsPASS now ships `mspass-dbverify` plus database verification methods for required attributes, schema/type checks, and normalization links, together with `dbclean` repair tooling and tests. The importing-data documentation recommends running the required and schema checks before processing.

More ambitious static validation of an entire arbitrary Python workflow would be a separate feature with different scope and acceptance criteria.
